airbyte-integrations/connectors/source-stripe/unit_tests/test_streams.py

```diff
@@ -104,72 +14,340 @@ def test_request_headers(accounts):
     assert headers["Stripe-Version"] == "2022-11-15"
 
 
-def test_lazy_sub_stream(requests_mock, invoice_line_items, invoices, stream_args):
-    # First initial request to parent stream
-    requests_mock.get(
-        "https://api.stripe.com/v1/invoices",
-        json={
-            "has_more": False,
-            "object": "list",
-            "url": "/v1/checkout/sessions",
-            "data": [
-                {
-                    "created": 1641038947,
-                    "customer": "cus_HezytZRkaQJC8W",
-                    "id": "in_1KD6OVIEn5WyEQxn9xuASHsD",
-                    "object": "invoice",
-                    "total": 1,
-                    "lines": {
-                        "data": [
-                            {
-                                "id": "il_1",
-                                "object": "line_item",
-                            },
-                            {
-                                "id": "il_2",
-                                "object": "line_item",
-                            },
-                        ],
-                        "has_more": True,
-                        "object": "list",
-                        "total_count": 3,
-                        "url": "/v1/invoices/in_1KD6OVIEn5WyEQxn9xuASHsD/lines",
+lazy_substream_test_suite = (
```

Comment:
I like this a lot! It feels like this is filling a big testing gap we have for streams where the dataset is hard/impossible to maintain. 

The PR addresses the case where the events stream records were not filtered properly. Should we test that using this kind of tests?

I would like to push the idea further: for each `stream_cls`,  do we want to have one test for incremental and one for full_refresh? The reason I'm asking is that the code path is very different for both. For example, `bank_accounts` relies on the `events` endpoint on incremental and on the customer else. One other crazy particularity with stripe is the pagination given `StripeLazySubStream` (for stream `invoice_line_items` for example) which seems to be a very specific and interesting case to document/test

Reply:
@maxi297 all these cases are already covered. 
- [Full refresh + pagination for Invoice Line Items (StripeLazySubStream)](https://github.com/airbytehq/airbyte/pull/32146/files#diff-426cbfecfffba519042f061d9fd04f2c94d90a82280251bebf29350326d1aaf4R18)
- [Full refresh + pagination for Subscription Items (StripeLazySubstream)](https://github.com/airbytehq/airbyte/pull/32146/files#diff-426cbfecfffba519042f061d9fd04f2c94d90a82280251bebf29350326d1aaf4R73)
- [Full refresh + pagination for Bank Accounts (UpdatedCursorIncrementalStripeLazySubstream)](https://github.com/airbytehq/airbyte/pull/32146/files#diff-426cbfecfffba519042f061d9fd04f2c94d90a82280251bebf29350326d1aaf4R128)
- [Full refresh + pagination for Application Fees Refunds (UpdatedCursorIncrementalStripeLazySubstream)](https://github.com/airbytehq/airbyte/pull/32146/files#diff-426cbfecfffba519042f061d9fd04f2c94d90a82280251bebf29350326d1aaf4R185)
- [Incremental for Bank Accounts (UpdatedCursorIncrementalStripeLazySubstream)](https://github.com/airbytehq/airbyte/pull/32146/files#diff-426cbfecfffba519042f061d9fd04f2c94d90a82280251bebf29350326d1aaf4R240)
- [Incremental for Application Fees Refunds (UpdatedCursorIncrementalStripeLazySubStream)](https://github.com/airbytehq/airbyte/pull/32146/files#diff-426cbfecfffba519042f061d9fd04f2c94d90a82280251bebf29350326d1aaf4R273)

Each of this cases tests:
 - Request args. The test would fail if the URL param values do not match those in the expected URL
 - Record filtering - the number of actual records should be the same as the number of expected records and it is not always the number of records in the mocked response
 - Cursor value - it is populated in the expected records, so if missing, the test would fail

Reply:
Ok, so we test per python classes and not per Stripe stream. I think that can be enough for now. What I'm worried about is that I don't see a safety net for a couple of changes you've done. For example:
* If we have an event that is not filtered properly
* If a dev removes `parent=expand_items=["data.sources"]` on bank_accounts

Do you think it's worth testing that?

Reply:
@maxi297 what do you mean by event that is not filtered properly? Did we have such an issue?
Regarding your second concern - I do think we do not have to cover this as we no more implement a class per stream  and this is an input option for instantiating a class

Reply:
I think we are looking at the same lines of code with two different perspectives that are not mutually exclusive but do not produce the same lines of code for the tests. 

>  I do think we do not have to cover this as we no more implement a class per stream and this is an input option for instantiating a class

This is true if we check the interface of the classes in isolation. So if we check the tests for `UpdatedCursorIncrementalStripeLazySubStream`, this case is covered as you mention; Unit tests are checking that just fine. Does that mean the source is working as we would like it to work? The answer is "no" as the code was failing in production and the stream BankAccounts was missing records in production. Tomorrow, I could remove [this line](https://github.com/airbytehq/airbyte/pull/32146/files#diff-e3651091c66723be70e9ab4b702a078c07f0e18fdbf895bba1b287a161576f3fR400) inadvertently and I would have no safety net to tell my I did something wrong because this is not caught on the unit test level or the CATs level. Hence, it feels like we need another safety net.

> what do you mean by event that is not filtered properly? Did we have such an issue?

I think we has as the PR mentions `Data not filtered when using the event-based incremental sync mode`. There was indeed [a change](https://github.com/airbytehq/airbyte/pull/32146/files#diff-bc939ec7f5dc075809864e54aca658012af002c91181af90b6b058de17bc4205R53-R54) made on filtering on the event stream. By looking more closely at the different cases, I see that [this entry](https://github.com/airbytehq/airbyte/pull/32146/files#diff-426cbfecfffba519042f061d9fd04f2c94d90a82280251bebf29350326d1aaf4R255-R262) would be filtered so the case is covered. What I fear is that the purpose of the test is not clear (it tests many things like that it perform the right query for incremental, filtering works, etc...) and as it isn't explicit, I can remove lines 255 to 262 from the test and it would still pass and still be a valid test. It's very hard to be too explicit but things are often too implicit.

As a more generic comments, tests have many many goals. We are pretty good in using tests for validation explicitly written code in a specific project. This is good as it ensure that all our small units of code work as expected. However, I think we should use tests for other reasons:
* Ensuring the integration between our component works as expected (hence the first point)
* Documenting the code (hence my second point)

Reply:
@maxi297 tests updated: Now we are testing not classes, but source streams. I have also split the generic test into three separate tests to verify that 
a. cursor value is populated,
b. data is filtered
c. data is expanded
for both sync modes (if applicable)

The PR has also been split into two different